src/App.js

```diff
@@ -1,15 +1,135 @@
-import React from 'react';
+import React, { Component } from 'react';
+import TodoList from './components/TodoList/TodoList';
+import getTodosWithUsers from './dataMappers';
 import './App.css';
 
-import todos from './api/todos';
-import users from './api/users';
+const API_URL = 'https://jsonplaceholder.typicode.com/';
 
-function App() {
-  return (
-    <div className="App">
-      <h1>Dynamic list of todos</h1>
-    </div>
-  );
+const getData = dataName => (
+  fetch(`${API_URL}${dataName}`)
+    .then(response => response.json())
+);
+
+class App extends Component {
+  state = {
+    todosList: [],
+    sortedTodosList: [],
+    isLoaded: false,
+    isLoading: false,
+    isError: false,
+    buttonText: 'Load',
+  }
+
+  loadDataFromServer = () => {
+    this.setState({
+      buttonText: 'loading...',
+      isLoading: true,
+    });
+
+    Promise.all([
+      getData('todos'),
+      getData('users'),
+    ])
+      .then(([todos, users]) => this.setState({
+        todosList: getTodosWithUsers(todos, users),
+        sortedTodosList: getTodosWithUsers(todos, users),
+        isLoaded: true,
+        isLoading: false,
+      }))
+      .catch(() => this.setState({
+        buttonText: 'try again',
+        isError: true,
+        isLoading: false,
+      }));
+  }
+
+  resetList = () => {
+    this.setState(prevState => ({
+      sortedTodosList: [...prevState.todosList],
+    }));
+  };
+
+  sortByTitle = () => {
+    this.setState(prevState => ({
+      sortedTodosList: [...prevState.todosList]
+        .sort((todo1, todo2) => todo1.title.localeCompare(todo2.title)),
+    }));
+  };
+
+  sortByName = () => {
+    this.setState(prevState => ({
+      sortedTodosList: [...prevState.todosList]
+        .sort((todo1, todo2) => todo1.user.name.localeCompare(todo2.user.name)),
+    }));
+  };
+
+  sortCompleted = () => {
+    this.setState(prevState => ({
+      sortedTodosList: [
+        ...prevState.todosList.filter(todo => todo.completed),
+        ...prevState.todosList.filter(todo => !todo.completed),
```

Comment:
instead of filtering twice create a variable, assign filtering result and then assign the value of this variable to the keys

Reply:
take this into consideration for other cases in this project

Reply:
we discussed this issue